File: acquisition/cadastre_clip_sda.py
#!/usr/bin/env python3
"""Clip d'un cadastre sur-capturé à la frontière municipale — source SDA EN MASSE.

Différence avec cadastre_clip.py (Overpass, 1 appel/ville) : la frontière vient
d'un INDEX LOCAL des municipalités SDA officielles (MRNF, 1/20k, EPSG:4326),
construit une seule fois pour toute la province. Aucun appel réseau par ville.

Résolution slug -> polygone SDA (slug = nom muni normalisé NFD sans accents/tirets) :
  1. norm(slug) == norm(MUS_NM_MUN) (exact)
  2. strip progressif des suffixes '--<mrc>' / '--2' (homonymes encodent leur MRC)
  3. homonymes -> désambiguïsation par MRC dans le suffixe, sinon par proximité
     spatiale (polygone SDA le plus proche du centroïde du cadastre)
  4. fallback index rôle MAMH (code géo) + petites alias connues
Clip par centroïde (representative_point + prep), anti-invention (suppression seule).

Usage module :
  from acquisition.cadastre_clip_sda import SDAIndex, clip_slug
  idx = SDAIndex("/path/qc-municipalites.geojson")           # + idx.attach_role_index()
  res = clip_slug(idx, slug, in_path, out_path, role_path)   # dict stats

Usage CLI :
  cadastre_clip_sda.py <slug> --boundaries qc-municipalites.geojson \
      [--in GEOJSON] [--out GEOJSON] [--role PARQUET]
"""
import json
import os
import re
import sys
import unicodedata
import argparse
from typing import Dict, List, Optional, Tuple

from shapely.geometry import shape, Point
from shapely.prepared import prep


# ---------------------------------------------------------------------------
# Normalisation slug (alignée sur la convention cadastre : apostrophes SUPPRIMÉES)
# ---------------------------------------------------------------------------

def norm(name: str) -> str:
    """Slug canonique : NFD sans accents, minuscule, apostrophes supprimées,
    tout non-alphanum -> tiret unique. 'Baie-D'Urfé' -> 'baie-durfe'."""
    nfd = unicodedata.normalize("NFD", name or "")
    a = "".join(c for c in nfd if unicodedata.category(c) != "Mn").lower()
    a = a.replace("'", "").replace("’", "").replace("`", "")
    a = re.sub(r"[^a-z0-9]+", "-", a)
    a = re.sub(r"-+", "-", a).strip("-")
    return a


# Alias slug cadastre -> code géo SDA (cas où le nom cadastre diffère du nom SDA)
ALIAS_SLUG_TO_CODE = {
    "eeyou-istchee-james-bay": "99060",      # Eeyou Istchee Baie-James (EN->FR)
    "hatley-township-municipality": "45043",  # Hatley (municipalité de canton)
}


# ---------------------------------------------------------------------------
# Index SDA
# ---------------------------------------------------------------------------

class SDAIndex:
    """Index municipalités SDA : norm(nom) -> [(code, nom, mrc_norm, geom, prep)]."""

    def __init__(self, boundaries_path: str):
        g = json.load(open(boundaries_path))
        self.by_name: Dict[str, List[dict]] = {}
        self.by_code: Dict[str, dict] = {}
        for f in g["features"]:
            p = f.get("properties") or {}
            code = (p.get("MUS_CO_GEO") or p.get("code") or "").strip()
            nm = p.get("MUS_NM_MUN") or p.get("name") or ""
            mrc = p.get("MUS_NM_MRC") or ""
            geom_dict = f.get("geometry")
            if not code or not nm or not geom_dict:
                continue
            try:
                geom = shape(geom_dict)
            except Exception:
                continue
            entry = {"code": code, "nom": nm, "mrc_norm": norm(mrc), "geom": geom}
            self.by_name.setdefault(norm(nm), []).append(entry)
            # by_code : si plusieurs polygones pour un code (territoires nordiques),
            # on fusionne en union pour avoir une frontière unique.
            if code in self.by_code:
                prev = self.by_code[code]
                try:
                    prev["geom"] = prev["geom"].union(geom)
                except Exception:
                    pass
            else:
                self.by_code[code] = dict(entry)
        self.role_index: Dict[str, Tuple[str, str]] = {}

    def attach_role_index(self, role_index_json: Optional[str] = None,
                          fetch: bool = False) -> int:
        """Charge un index rôle {norm(nom): (code, nom)} pour le fallback code.
        role_index_json : JSON pré-fetché ; fetch=True télécharge l'index MAMH."""
        if role_index_json and os.path.exists(role_index_json):
            raw = json.load(open(role_index_json))
            self.role_index = {k: tuple(v) for k, v in raw.items()}
        elif fetch:
            import urllib.request
            import csv as _csv
            import io as _io
            url = "https://donneesouvertes.affmunqc.net/role/indexRole2026.csv"
            content = urllib.request.urlopen(url, timeout=40).read().decode("utf-8-sig")
            reader = _csv.DictReader(_io.StringIO(content))
            for r in reader:
                cg = (r.get("code géographique") or "").strip()
                nm = (r.get("nom du territoire") or "").strip()
                if cg:
                    self.role_index[norm(nm)] = (cg, nm)
        return len(self.role_index)


# ---------------------------------------------------------------------------
# Résolution slug -> entrée(s) candidate(s)
# ---------------------------------------------------------------------------

def _split_slug(slug: str) -> List[str]:
    """Découpe un slug brut sur le séparateur '--' (avant normalisation, qui
    écraserait '--' en '-') et norme chaque segment. Le 1er segment = nom muni,
    les suivants = MRC / compteur ('2')."""
    raw = slug or ""
    segs = [seg for seg in re.split(r"-{2,}", raw)]
    return [norm(seg) for seg in segs]


def _candidates_for(idx: SDAIndex, slug: str) -> List[dict]:
    """Liste des entrées SDA candidates pour un slug (peut être >1 si homonyme)."""
    s = norm(slug)
    if s in idx.by_name:
        return idx.by_name[s]
    segs = _split_slug(slug)
    # essaie le 1er segment (nom muni) seul, puis des préfixes décroissants
    for i in range(len(segs), 0, -1):
        base = norm("-".join(segs[:i]))
        if base in idx.by_name:
            return idx.by_name[base]
    return []


def resolve_boundary(idx: SDAIndex, slug: str,
                     near: Optional[Tuple[float, float]] = None
                     ) -> Tuple[Optional[dict], str]:
    """Résout un slug cadastre vers UNE entrée SDA (geom). Retourne (entry, method).
    entry=None si aucun match. `near`=(lon,lat) centroïde cadastre pour
    désambiguïser les homonymes par proximité spatiale."""
    s = norm(slug)

    # 0) alias explicite
    if s in ALIAS_SLUG_TO_CODE:
        e = idx.by_code.get(ALIAS_SLUG_TO_CODE[s])
        if e:
            return e, "alias"

    cands = _candidates_for(idx, slug)

    if len(cands) == 1:
        return cands[0], "name"

    if len(cands) > 1:
        # désambig par MRC encodée dans le suffixe '--<mrc>'
        segs = _split_slug(slug)
        suffix_norm = norm("-".join(segs[1:])) if len(segs) > 1 else ""
        if suffix_norm:
            for c in cands:
                if c["mrc_norm"] and (c["mrc_norm"] == suffix_norm
                                      or c["mrc_norm"] in suffix_norm
                                      or suffix_norm in c["mrc_norm"]):
                    return c, "name+mrc"
        # sinon proximité spatiale
        if near is not None:
            pt = Point(near)
            cands = sorted(cands, key=lambda c: c["geom"].distance(pt))
            return cands[0], "name+proximity"
        return cands[0], "name+firstambig"

    # fallback index rôle (code géo) -> geom SDA
    if idx.role_index:
        ri = idx.role_index.get(s)
        if not ri:
            segs = _split_slug(slug)
            for i in range(len(segs), 0, -1):
                base = norm("-".join(segs[:i]))
                if base in idx.role_index:
                    ri = idx.role_index[base]
                    break
        if ri:
            e = idx.by_code.get(ri[0])
            if e:
                return e, "role-index"

    return None, "no-match"


# ---------------------------------------------------------------------------
# Clip
# ---------------------------------------------------------------------------

def clip_slug(idx: SDAIndex, slug: str, in_path: str, out_path: str,
              role_path: Optional[str] = None) -> dict:
    """Clippe le cadastre `slug` à sa frontière SDA. Retourne stats + diagnostic.
    Anti-invention : ne retient QUE les lots dont le centroïde tombe dans la
    frontière muni ; geom inchangée ; aucun ajout."""
    g = json.load(open(in_path))
    feats = g.get("features", [])
    n = len(feats)

    # centroïdes (representative_point) + centroïde global pour désambiguïsation
    cents: List[Optional[Point]] = []
    xs: List[float] = []
    ys: List[float] = []
    for f in feats:
        geom = f.get("geometry")
        if not geom:
            cents.append(None)
            continue
        try:
            c = shape(geom).representative_point()
        except Exception:
            cents.append(None)
            continue
        cents.append(c)
        xs.append(c.x)
        ys.append(c.y)
    near = (sum(xs) / len(xs), sum(ys) / len(ys)) if xs else None

    entry, method = resolve_boundary(idx, slug, near=near)
    if entry is None:
        return {"slug": slug, "before": n, "after": 0, "retained_pct": 0.0,
                "boundary_match": False, "resolve_method": method,
                "out": None}

    muni = entry["geom"]
    pm = prep(muni)
    kept = []
    kept_nolots = set()
    for f, c in zip(feats, cents):
        if c is not None and pm.contains(c):
            kept.append(f)
            nl = str((f.get("properties") or {}).get("NO_LOT") or "").replace(" ", "")
            if nl:
                kept_nolots.add(nl)

    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    with open(out_path, "w") as fh:
        json.dump({"type": "FeatureCollection", "features": kept}, fh)

    res = {
        "slug": slug,
        "before": n,
        "after": len(kept),
        "retained_pct": round(100 * len(kept) / n, 1) if n else 0.0,
        "boundary_match": True,
        "resolve_method": method,
        "sda_code": entry["code"],
        "sda_nom": entry["nom"],
        "muni_area_deg2": round(muni.area, 5),
        "out": out_path,
    }

    if role_path and os.path.exists(role_path):
        import pyarrow.parquet as pq
        t = pq.read_table(role_path)
        cols = set(t.column_names)
        if "NO_LOT" in cols and "role_usage_cubf" in cols:
            NO = t.column("NO_LOT").to_pylist()
            CU = t.column("role_usage_cubf").to_pylist()
            matched = set((NO[i] or "").replace(" ", "")
                          for i in range(len(NO)) if CU[i] is not None)
            inter = kept_nolots & matched
            res["join_before"] = round(100 * len(matched) / n, 1) if n else 0.0
            res["join_after"] = round(100 * len(inter) / len(kept), 1) if kept else 0.0
            res["matched_coverage"] = (round(100 * len(inter) / len(matched), 1)
                                       if matched else 0.0)
    return res


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("slug")
    ap.add_argument("--boundaries", required=True,
                    help="GeoJSON municipalités SDA (EPSG:4326)")
    ap.add_argument("--role-index", help="JSON index rôle {norm:(code,nom)} (fallback)")
    ap.add_argument("--fetch-role-index", action="store_true")
    ap.add_argument("--in", dest="in_path")
    ap.add_argument("--out", dest="out_path")
    ap.add_argument("--role", dest="role_path")
    a = ap.parse_args()
    idx = SDAIndex(a.boundaries)
    if a.role_index or a.fetch_role_index:
        idx.attach_role_index(a.role_index, fetch=a.fetch_role_index)
    in_path = a.in_path or "/tmp/lots/%s.geojson" % a.slug
    out_path = a.out_path or "/tmp/clipped/%s.geojson" % a.slug
    r = clip_slug(idx, a.slug, in_path, out_path, a.role_path)
    print(json.dumps(r, ensure_ascii=False))

File: acquisition/src/cadastre-clip-sda.ts
/**
 * cadastre-clip-sda.ts — Clip d'un cadastre sur-capturé à la frontière
 * municipale SDA (index local, 0 appel réseau/ville).
 *
 * Port fidèle de `acquisition/cadastre_clip_sda.py`. Résolution slug -> polygone
 * SDA par `norm()` (NFD sans accents, apostrophes supprimées), désambiguïsation
 * des homonymes par MRC dans le suffixe `--<mrc>` puis proximité spatiale, plus
 * un fallback index rôle MAMH (code géo). Clip par centroïde (representative
 * point + strict point-in-polygon), anti-invention (suppression seule).
 */
import { readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

import type { Feature, Geometry, Polygon, MultiPolygon, Position } from "geojson";
import distance from "@turf/distance";
import { representativePoint, strictPointInPolygon } from "./lib/geo.js";

/**
 * Slug canonique : NFD sans accents (catégorie Mn retirée), minuscule,
 * apostrophes supprimées, tout non-alphanum -> tiret unique.
 * "Baie-D'Urfé" -> "baie-durfe". IDENTIQUE à `cadastre_clip_sda.norm`.
 */
export function norm(name: string): string {
  const nfd = (name ?? "").normalize("NFD");
  let a = nfd
    .replace(/\p{Mn}/gu, "") // strip combining accents (Unicode category Mn)
    .toLowerCase();
  a = a.replace(/['’`]/g, "");
  a = a.replace(/[^a-z0-9]+/g, "-");
  a = a.replace(/-+/g, "-").replace(/^-+|-+$/g, "");
  return a;
}

export const ALIAS_SLUG_TO_CODE: Record<string, string> = {
  "eeyou-istchee-james-bay": "99060",
  "hatley-township-municipality": "45043",
};

export interface SdaEntry {
  code: string;
  nom: string;
  mrc_norm: string;
  geom: Polygon | MultiPolygon;
}

type GeoJSONFeature = Feature<Geometry, Record<string, unknown>>;

/** Union de deux (multi)polygones en MultiPolygon (territoires nordiques). */
function unionPolys(
  a: Polygon | MultiPolygon,
  b: Polygon | MultiPolygon,
): MultiPolygon {
  const toParts = (g: Polygon | MultiPolygon): Position[][][] =>
    g.type === "Polygon" ? [g.coordinates] : g.coordinates;
  return { type: "MultiPolygon", coordinates: [...toParts(a), ...toParts(b)] };
}

export class SDAIndex {
  byName: Map<string, SdaEntry[]> = new Map();
  byCode: Map<string, SdaEntry> = new Map();
  roleIndex: Map<string, [string, string]> = new Map();

  constructor(boundariesPath: string) {
    const g = JSON.parse(readFileSync(boundariesPath, "utf8")) as {
      features: GeoJSONFeature[];
    };
    for (const f of g.features) {
      const p = f.properties ?? {};
      const code = String(p["MUS_CO_GEO"] ?? p["code"] ?? "").trim();
      const nm = String(p["MUS_NM_MUN"] ?? p["name"] ?? "");
      const mrc = String(p["MUS_NM_MRC"] ?? "");
      const geomDict = f.geometry;
      if (!code || !nm || !geomDict) continue;
      if (geomDict.type !== "Polygon" && geomDict.type !== "MultiPolygon") continue;
      const entry: SdaEntry = {
        code,
        nom: nm,
        mrc_norm: norm(mrc),
        geom: geomDict as Polygon | MultiPolygon,
      };
      const key = norm(nm);
      if (!this.byName.has(key)) this.byName.set(key, []);
      this.byName.get(key)!.push(entry);
      const prev = this.byCode.get(code);
      if (prev) {
        try {
          prev.geom = unionPolys(prev.geom, entry.geom);
        } catch {
          /* keep prev */
        }
      } else {
        this.byCode.set(code, { ...entry });
      }
    }
  }

  /** Charge un index rôle {norm(nom): [code, nom]} (JSON pré-fetché). */
  attachRoleIndex(roleIndexJson?: string): number {
    if (roleIndexJson) {
      const raw = JSON.parse(readFileSync(roleIndexJson, "utf8")) as Record<
        string,
        [string, string]
      >;
      this.roleIndex = new Map(Object.entries(raw).map(([k, v]) => [k, v]));
    }
    return this.roleIndex.size;
  }
}

/** Découpe un slug brut sur '--' (avant normalisation) et norme chaque segment. */
function splitSlug(slug: string): string[] {
  return (slug ?? "").split(/-{2,}/).map((seg) => norm(seg));
}

function candidatesFor(idx: SDAIndex, slug: string): SdaEntry[] {
  const s = norm(slug);
  if (idx.byName.has(s)) return idx.byName.get(s)!;
  const segs = splitSlug(slug);
  for (let i = segs.length; i > 0; i--) {
    const base = norm(segs.slice(0, i).join("-"));
    if (idx.byName.has(base)) return idx.byName.get(base)!;
  }
  return [];
}

/** Distance planaire approx (degrés) entre point et entrée — pour le tri de
 * proximité, équivalent monotone à shapely `geom.distance(pt)` (centroïde). */
function distToEntry(entry: SdaEntry, near: Position): number {
  const rp = representativePoint(entry.geom);
  const c = rp ?? near;
  // turf distance (km) is monotone in planar distance for ranking purposes.
  return distance(c as [number, number], near as [number, number]);
}

export function resolveBoundary(
  idx: SDAIndex,
  slug: string,
  near: Position | null = null,
): [SdaEntry | null, string] {
  const s = norm(slug);
  if (ALIAS_SLUG_TO_CODE[s]) {
    const e = idx.byCode.get(ALIAS_SLUG_TO_CODE[s]!);
    if (e) return [e, "alias"];
  }
  const cands = candidatesFor(idx, slug);
  if (cands.length === 1) return [cands[0]!, "name"];
  if (cands.length > 1) {
    const segs = splitSlug(slug);
    const suffixNorm = segs.length > 1 ? norm(segs.slice(1).join("-")) : "";
    if (suffixNorm) {
      for (const c of cands) {
        if (
          c.mrc_norm &&
          (c.mrc_norm === suffixNorm ||
            c.mrc_norm.includes(suffixNorm) ||
            suffixNorm.includes(c.mrc_norm))
        ) {
          return [c, "name+mrc"];
        }
      }
    }
    if (near !== null) {
      const sorted = [...cands].sort(
        (a, b) => distToEntry(a, near) - distToEntry(b, near),
      );
      return [sorted[0]!, "name+proximity"];
    }
    return [cands[0]!, "name+firstambig"];
  }
  // fallback role index
  if (idx.roleIndex.size) {
    let ri = idx.roleIndex.get(s);
    if (!ri) {
      const segs = splitSlug(slug);
      for (let i = segs.length; i > 0; i--) {
        const base = norm(segs.slice(0, i).join("-"));
        if (idx.roleIndex.has(base)) {
          ri = idx.roleIndex.get(base);
          break;
        }
      }
    }
    if (ri) {
      const e = idx.byCode.get(ri[0]);
      if (e) return [e, "role-index"];
    }
  }
  return [null, "no-match"];
}

export interface ClipResult {
  slug: string;
  before: number;
  after: number;
  retained_pct: number;
  boundary_match: boolean;
  resolve_method: string;
  sda_code?: string;
  sda_nom?: string;
  muni_area_deg2?: number;
  out: string | null;
  join_before?: number;
  join_after?: number;
  matched_coverage?: number;
}

/** Aire planaire (deg²) d'un (multi)polygone par la formule du lacet — équivaut
 * à shapely `.area` en degrés (mesure de diagnostic uniquement). */
function planarAreaDeg2(geom: Polygon | MultiPolygon): number {
  const ringArea = (ring: Position[]): number => {
    let a = 0;
    for (let i = 0; i < ring.length - 1; i++) {
      a += ring[i]![0]! * ring[i + 1]![1]! - ring[i + 1]![0]! * ring[i]![1]!;
    }
    return Math.abs(a) / 2;
  };
  const polyArea = (poly: Position[][]): number => {
    if (poly.length === 0) return 0;
    let a = ringArea(poly[0]!);
    for (let i = 1; i < poly.length; i++) a -= ringArea(poly[i]!);
    return a;
  };
  if (geom.type === "Polygon") return polyArea(geom.coordinates);
  return geom.coordinates.reduce((s, p) => s + polyArea(p), 0);
}

/**
 * Charge la cadastre, résout la frontière SDA, ne retient QUE les lots dont le
 * centroïde (representative point) tombe strictement dans la frontière.
 * Si `roleParquetRows` est fourni (lu via le reader parquet), recalcule le
 * join% post-clip — IDENTIQUE à la branche pyarrow du .py.
 */
export function clipSlug(
  idx: SDAIndex,
  slug: string,
  inPath: string,
  outPath: string,
  roleParquetRows?: { NO_LOT?: unknown; role_usage_cubf?: unknown }[] | null,
): ClipResult {
  const g = JSON.parse(readFileSync(inPath, "utf8")) as {
    features: GeoJSONFeature[];
  };
  const feats = g.features ?? [];
  const n = feats.length;

  const cents: (Position | null)[] = [];
  const xs: number[] = [];
  const ys: number[] = [];
  for (const f of feats) {
    const c = representativePoint(f.geometry);
    cents.push(c);
    if (c) {
      xs.push(c[0]!);
      ys.push(c[1]!);
    }
  }
  const near: Position | null = xs.length
    ? [xs.reduce((a, b) => a + b, 0) / xs.length, ys.reduce((a, b) => a + b, 0) / ys.length]
    : null;

  const [entry, method] = resolveBoundary(idx, slug, near);
  if (entry === null) {
    return {
      slug,
      before: n,
      after: 0,
      retained_pct: 0,
      boundary_match: false,
      resolve_method: method,
      out: null,
    };
  }

  const muni = entry.geom;
  const kept: GeoJSONFeature[] = [];
  const keptNolots = new Set<string>();
  for (let i = 0; i < feats.length; i++) {
    const c = cents[i];
    if (c && strictPointInPolygon(c, muni)) {
      kept.push(feats[i]!);
      const nl = String((feats[i]!.properties ?? {})["NO_LOT"] ?? "").replace(/ /g, "");
      if (nl) keptNolots.add(nl);
    }
  }

  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, JSON.stringify({ type: "FeatureCollection", features: kept }));

  const res: ClipResult = {
    slug,
    before: n,
    after: kept.length,
    retained_pct: n ? round1((100 * kept.length) / n) : 0,
    boundary_match: true,
    resolve_method: method,
    sda_code: entry.code,
    sda_nom: entry.nom,
    muni_area_deg2: round5(planarAreaDeg2(muni)),
    out: outPath,
  };

  if (roleParquetRows && roleParquetRows.length) {
    const matched = new Set<string>();
    for (const r of roleParquetRows) {
      if (r.role_usage_cubf !== null && r.role_usage_cubf !== undefined) {
        matched.add(String(r.NO_LOT ?? "").replace(/ /g, ""));
      }
    }
    let inter = 0;
    for (const k of keptNolots) if (matched.has(k)) inter++;
    res.join_before = n ? round1((100 * matched.size) / n) : 0;
    res.join_after = kept.length ? round1((100 * inter) / kept.length) : 0;
    res.matched_coverage = matched.size ? round1((100 * inter) / matched.size) : 0;
  }
  return res;
}

function round1(x: number): number {
  return Math.round(x * 10) / 10;
}
function round5(x: number): number {
  return Math.round(x * 1e5) / 1e5;
}
